content/post/2019-01-20-what-is-production-grade.markdown

```diff
@@ -0,0 +1,270 @@
+---
+title: "What Is Production Grade?"
+date: 2019-01-20T15:35:03-08:00
+---
+
+I recently decided to try the Caddy webserver project's implementation
+of a webserver written in Golang.  It had many interesting things
+going for it, including automagic certificate renewal via ACME and
+Let's Encrypt, "plugins" written in a sane language, dramatically
+simplified configuration, and perhaps in one of the most boastful
+statements I've seen of any software recently, it claims to be
+"Production-Ready" (features page, right hand side).  Given that it
+took me the better part of a day to get a functional Caddy role for
+Ansible, I really question this claim.
+
+So that's what this article is about.  I'm putting down here what I
+think production grade software should be, and some of the problems
+with what software is out there.
+
+## What is Production?
+
+To answer the original question, we have to first look into what is
+production.  In the general sense, production is agreed to be the
+environment of a software system that is either directly or
+transitively involved in serving end-users.  These end-users can be
+internal or external to an organization, but the defining
+characteristic that makes someone an end user is that they can be
+unhappy about the status of a system and their unhappiness directly
+translates into a need to act for some team responsible for the
+system.  This is in contrast to dev or QA environments where
+unhappiness about the system state does not demand action.
+
+Put more simply and more directly: if it is important enough to wake
+some one up and put a keyboard in front of them, its part of
+production.
+
+With a reasonable definition of what production is, lets now look at
+what production grade means.
+
+## What is Production Grade?
+
+Like lumber, meat, and equipment, its not unreasonable to say that
+software comes in various grades.  In many cases these terms are
+legally protected.  For example the United States Department of
+Agriculture governs the use of specific terms for grades of meat.
+Similarly the Western Wood Products Association provides grading
+criteria for lumber.  It is interesting then that no such set of
+standards exists for software.  Anyone can claim at any time that any
+software is production grade, or is of sufficient quality to wake
+someone up at night.
+
+Its at this point that I'm going to clearly state that for the
+purposes of this article, I'm glossing over the complexity of software
+fitness for a particular purpose, such as life or safety critical
+systems.  These systems usually require the review and certification
+of a Professional Engineer, whose title and license are part of a
+regulated profession in most parts of the world.  I am not a
+Professional Engineer, and so I can't speak to the kinds of things
+that one would look for in certifying software.  I can only speak to
+what I look for in things that I allow to wake me up at night.
```

Comment:
The use of `things` in a meaning distinct from the use in the prior sentence is slightly confusing. Perhaps use `software` here instead?